Getting started: Developer section (#6081)
braze-inc/braze-docs: _docs/_developer_guide/platform_wide/getting_started/integration_overview.md

article_title: Integration Overview
page_order: 2
description: "This article provides a basic overview of the onboarding process."
platform:
  - iOS
  - Android
  - Web
  - React Native
  - Flutter
  - Cordova
  - Roku
  - Swift
  - Unity
---

# [![Braze Learning course]({% image_buster /assets/img/bl_icon2.png %})](https://learning.braze.com/sdk-integration-basics){: style="float:right;width:120px;border:0;" class="noimgborder"}Getting started: Integration overview

> This article provides a basic overview of the onboarding process.

![A venn diagram of four circles - discovery, integrate, quality assurance and maintain - centered around "time to value."][27]{: style="max-width:50%;float:right;margin-left:15px;"} 

As a technical resource, you'll empower your team by integrating Braze into your tech stack. Onboarding is broadly split up into four steps:
* [Discovery and planning](#discovery): Work with your team to align on scope, plan a structure for data and campaigns, and create an appropriate workspace structure. 
* [Integration](#integration): Execute on your plan by integrating the SDK and API, enabling messaging channels, and setting up data import and export. 

> * [Quality Assurance](#qa): Ensure that the loop of data and messaging between the Braze platform and your app or site is working as expected.

Reviewer: ```suggestion
* [Quality Assurance](#qa): Confirm that the loop of data and messaging between the Braze platform and your app or site is working as expected.
```

> * [Maintenance](#maintenance): Once you've passed off Braze to your Marketing team, you'll continue to ensure everything continues to run smoothly.

Reviewer: ```suggestion
* [Maintenance](#maintenance): After you've passed off Braze to your marketing team, you'll continue to make sure everything continues to run smoothly.
```

<br>
{% alert tip %}
We recognize that every organization has its distinct needs, and Braze is built to cater to a diverse range of customization options that can be tailored to your specific requirements. Integration times will vary based on your use case. 
{% endalert %}

## Discovery and planning {#discovery}
During this phase, you will work with your team to scope onboarding tasks and ensure all stakeholders are aligned on a common goal. 

> Your team will perform end-to-end planning of your use cases to ensure everything can be built as expected, with the correct data available to do so. This phase includes your project lead, CRM lead, front and back-end Engineering, product owners, and marketers. 

Reviewer: ```suggestion
Your team will perform end-to-end planning of your use cases to make sure everything can be built as expected, with the correct data available to do so. This phase includes your project lead, CRM lead, front and back-end engineering, product owners, and marketers. 
```

{% endalert %}

![Icons for different channels, such as email, shopping cart, images, geolocation, and so on.][28]{: style="max-width:40%;float:right;margin-left:15px;"} 

### Campaign planning

Your CRM team will plan out the messaging use cases that you'll launch in the near future. This includes the:
* [Channel][1] (for example, push notifications or in-app messages)
* [Delivery method][2] (for example, scheduled delivery or action-based delivery)
* [Target audience][3]
* [Success metrics][4]

For example, a new customer campaign might be: an email sent daily at 10 am to a segment of customers who logged their first session yesterday. The conversion event (the success metric) is logging a session.

<br>
{% alert important %}
Integration cannot begin until the campaign planning step is complete. This step will determine what parts and pieces of Braze need to be configured during the integration phase.
{% endalert %}

### Creating data requirements
Then, your CRM team should define what data is required to launch the campaigns they have planned, creating data requirements. 

> Many common types of user attributes, such as name, email, date of birth, country, etc. are automatically tracked once the Braze SDK is integrated. Other types of data will need to be defined as custom data.

Reviewer: ```suggestion
Many common types of user attributes, such as name, email, date of birth, country, and similar are automatically tracked once the Braze SDK is integrated. Other types of data will need to be defined as custom data.
```

As a developer, you'll work with your team to define what additional, custom data makes sense to track. Your custom data will impact how your user base will be classified and segmented. You will set up an event taxonomy across your growth stack, structuring your data so that it is compatible with your systems as it moves in and out of Braze.

{% alert tip %}
Keep data nomenclature consistent across tools. For example, your data warehouse may record "purchase limited time offer" in a particular way. You will need to decide if a custom event in Braze is needed to match this format.
{% endalert %}

Learn more about [automatically collected data and custom data][5].

### Customizations planning

> Talk to your marketers about their desired customizations. For example, do you want to implement the default Braze Content Cards? Do you want to slightly tweak their look and feel to match your brand guidelines? Do you want to develop an entirely new UI for a component and have Braze track its analytics? Different levels of customization require different levels of scope; see our [customization overview for more information][6].

Reviewer 2: Note on this as well, here we don't capitalize "customization"but in the platform overview article, we do.

```suggestion
Talk to your marketers about their desired customizations. For example, do you want to implement the default Braze Content Cards? Do you want to slightly tweak their look and feel to match your brand guidelines? Do you want to develop an entirely new UI for a component and have Braze track its analytics? Different levels of customization require different levels of scope; see our [customization overview][6] for more information.
```

### Getting dashboard access
The Braze dashboard is our web UI interface. Marketers will use the dashboard to do their job and create content. Developers use the dashboard to manage settings for integrating apps, such as API keys and push notification credentials.

Your team administrator should add you (and all other team members who need access to Braze) as users on your dashboard.

### Workspaces and API keys

> Your team administrator will also create different [workspaces]({{site.baseurl}}/developer_guide/platform_wide/getting_started/app_group_configuration/). Workspaces group your data—users, segments, API keys—into one location. As a best practice, we suggest only putting different versions of the same or very similar apps together under one workspace. 

Reviewer: The link opens to a 404

Reviewer 2: Working on review commenting now, will add the correct link as a suggestion

> Your team administrator will also create different [workspaces]({{site.baseurl}}/developer_guide/platform_wide/getting_started/app_group_configuration/). Workspaces group your data—users, segments, API keys—into one location. As a best practice, we suggest only putting different versions of the same or very similar apps together under one workspace. 

Reviewer 2: Correct link
```suggestion
Your team administrator will also create different [workspaces]({{site.baseurl}}/user_guide/administrative/app_settings/workspaces/). Workspaces group your data—users, segments, API keys—into one location. As a best practice, we suggest only putting different versions of the same or very similar apps together under one workspace. 
```

> Importantly, workspaces provide API keys for multiple platforms (iOS, Android, etc.). You'll use the correlated API keys to associate SDK data with a particular workspace. Navigate to your workspaces to access the API key for each of your apps. Ensure each API key has the correct permissions to perform the work you've scoped. See the [API provisioning article for details.][7]

Reviewer: ```suggestion
Importantly, workspaces provide API keys for multiple platforms (such as iOS and Android). You'll use the correlated API keys to associate SDK data with a particular workspace. Navigate to your workspaces to access the API key for each of your apps. Ensure each API key has the correct permissions to perform the work you've scoped. See the [API provisioning article for details.][7]
```

> Importantly, workspaces provide API keys for multiple platforms (such as iOS and Android). You'll use the correlated API keys to associate SDK data with a particular workspace. Navigate to your workspaces to access the API key for each of your apps. Ensure each API key has the correct permissions to perform the work you've scoped. See the [API provisioning article for details.][7]

Reviewer: ```suggestion
Importantly, workspaces provide API keys for multiple platforms (such as iOS and Android). You'll use the correlated API keys to associate SDK data with a particular workspace. Navigate to your workspaces to access the API key for each of your apps. Make sure each API key has the correct permissions to perform the work you've scoped. See the [API provisioning article for details.][7]
```

> Importantly, workspaces provide API keys for multiple platforms (such as iOS and Android). You'll use the correlated API keys to associate SDK data with a particular workspace. Navigate to your workspaces to access the API key for each of your apps. Ensure each API key has the correct permissions to perform the work you've scoped. See the [API provisioning article for details.][7]

Reviewer 2: Another minor link standard violation 👀 
```suggestion
Importantly, workspaces provide API keys for multiple platforms (such as iOS and Android). You'll use the correlated API keys to associate SDK data with a particular workspace. Navigate to your workspaces to access the API key for each of your apps. Ensure each API key has the correct permissions to perform the work you've scoped. See the [API provisioning article][7] for details.
```

{% alert important %}

> It's important that you set up different environments for development and production. Setting up a test environment will ensure you won't spend actual money during onboarding and QA. To create a testing environment, set up a testing workspace and be sure to use its API key so that you aren't populating your production workspace with test data.

Reviewer: ```suggestion
It's important that you set up different environments for development and production. Setting up a test environment will prevent you from spending actual money during onboarding and QA. To create a testing environment, set up a testing workspace and be sure to use its API key so that you aren't populating your production workspace with test data.
```

If you are performing side-by-side integration with a CDP, you will map the calls from your CDP's SDK to the Braze SDK. Essentially, you will:
* Map identifying calls to `changeUser` ([Android][11], [iOS][12], [web][13]) and set attributes.
* Map data flush calls to `requestImmediateDataFlush` ([Android][14], [iOS][15], [web][16]).
* Log custom events or purchases.

Example integrations between the Braze SDK and your CDP of choice might be available, depending on which platform you've chosen. See our [list of CDP technology partners][8] for more information. 

### Braze SDK integration 
The Braze SDK provides two critical pieces of functionality: it collects and syncs user data into a consolidated user profile, and powers messaging channels such as push notifications, in-app messages, and Content Cards. 

{% alert tip %} 
When fully integrated with your app or site, the Braze SDK offers a completely-realized level of marketing sophistication. If you defer integrating the Braze SDK, some of the functionality described in the documentation will not be available.
{% endalert %}

During SDK implementation, you will:
* Write SDK integration code for each platform you want to support.

> * Activate the messaging channels for each platform, ensuring that the Braze SDK tracks the data from your interactions with your customers across email, SMS, push notifications, etc.

Reviewer: ```suggestion
* Activate the messaging channels for each platform, ensuring that the Braze SDK tracks the data from your interactions with your customers across email, SMS, push notifications, and other channels.
```

### Using the Braze API
You will use our REST API for different tasks at different points throughout your time using Braze. The Braze API is useful for: 
1. Importing historical data; and
2. Continuous updates that aren’t triggered in Braze. For example, a user profile upgrades to VIP without them logging into an app, so the API needs to communicate this info to Braze.

Get started with the [Braze API.][9]

{% alert important %}
While using the API, ensure you batch your requests and only send delta values. Braze re-writes every attribute that is sent. Do not update any custom attribute if its value has not changed.
{% endalert %}

### Setting up product analytics
Braze is all about data. Data in Braze is stored on the user profile. 

Data points are a structure by which you ensure you're capturing the right data for your marketers, not just "any" data you can possibly vacuum up. Familiarize yourself with [data points][10].

### Migrating legacy user data

> You can use [Braze's Track Users endpoint][17] to migrate historical data that was recorded outside of Braze. Examples of commonly imported data include push tokens and past purchases. This endpoint can be used for one-off imports or regular batch updates. 

Reviewer 2: This is historically how we format linking to this endpoint 🤔 cc: @lydia-xie 
```suggestion
You can use the Braze [/users/track endpoint][17] to migrate historical data that was recorded outside of Braze. Examples of commonly imported data include push tokens and past purchases. This endpoint can be used for one-off imports or regular batch updates. 
```

{% endalert %}

### Other tools
Based on your use case, there may be other tools you need to set up. For example, you might need to configure a tool like [geofences][24] to realize your user stories. We have found that customers who have the ability to set up these additional tools after completing the essential integration steps are most successful.

## Quality assurance {#qa}
As you execute your integration, you'll provide quality assurance to make sure everything you're setting up is working as expected. This QA falls into two general categories: data ingestion and message channels.

{% alert important %}
Make sure your production and testing environments are set up before beginning QA.
{% endalert %}

| **QA data ingestion**  | **QA messaging**                                              |
|---------------------------|---------------------------------------------------------------|

> | You'll QA the way data is ingested, stored, and exported. | You'll ensure that your messages are being sent correctly to your users and everything looks excellent. |

Reviewer: ```suggestion
| You'll QA the way data is ingested, stored, and exported. | You'll make sure that your messages are being sent correctly to your users and everything looks excellent. |
```

| Run tests to confirm data is stored properly. | Create segments of users. |

> | Ensure session data is correctly attributed to the intended workspace within Braze. | Launch campaigns and Canvases successfully. |

Reviewer: ```suggestion
| Confirm session data is correctly attributed to the intended workspace within Braze. | Launch campaigns and Canvases successfully. |
```

> | Ensure session starts and ends are being recorded. | Ensure the correct campaigns are being shown to the correct user segments. |

Reviewer: ```suggestion
| Confirm session starts and ends are being recorded. | Confirm the correct campaigns are being shown to the correct user segments. |
```

> | Ensure user attribute information is correctly recorded against user profiles. | Ensure that push tokens are correctly being registered. |

Reviewer: ```suggestion
| Confirm user attribute information is correctly recorded against user profiles. | Confirm that push tokens are correctly being registered. |
```

> | Test that custom data is being correctly recorded against user profiles. | Ensure that push tokens are correctly removed. |

Reviewer: ```suggestion
| Test that custom data is being correctly recorded against user profiles. | Confirm that push tokens are correctly removed. |
```

| Create anonymous user profiles. | Test that push campaigns are correctly sending to devices and engagement is logged. |

> | Ensure that anonymous user profiles become known user profiles when the `changeUser()` method is called. | Test that in-app messages are delivered and metrics logged. |

Reviewer: ```suggestion
| Confirm that anonymous user profiles become known user profiles when the `changeUser()` method is called. | Test that in-app messages are delivered and metrics logged. |
```

|                           | Test that Content Cards are delivered and metrics logged. |
|                           | Facilitate Connected Content (for example, AccuWeather). |

> |                           | Ensure all message channel integrations are working together properly. |

Reviewer: ```suggestion
|                           | Confirm all message channel integrations are working together properly. |
```

{: .reset-td-br-1 .reset-td-br-2 .reset-td-br-3 .reset-td-br-4}

### Passing Braze off to marketers
Once you have integrated your platform or site, you will want to involve your Marketing team to pass ownership of the platform to them. This process looks different at every company, but might include the following:
* Composing complex [Liquid logic][25]
* Help facilitating [email IP warming][26]
* Making sure other stakeholders understand the kind of data being tracked

### Develop for the future
Have you ever inherited a codebase and had no clue what the initial developer was thinking? Worse, have you ever written code, understood it completely, and then felt completely baffled when you came back to it a year later? 

> When onboarding Braze, the collective decisions you make concerning data, user profiles, what integrations were and were not in scope, how customizations are supposed to work, etc., will feel fresh in your mind and therefore obvious. When your team wants to expand Braze or when other technical resources are assigned to your Braze project, this information will be obscure.

Reviewer: ```suggestion
When onboarding Braze, the collective decisions you make concerning data, user profiles, what integrations were and were not in scope, how customizations are supposed to work, and more, will feel fresh in your mind and therefore obvious. When your team wants to expand Braze or when other technical resources are assigned to your Braze project, this information will be obscure.
```

Create a resource to cement the information you learned during your technical overview sessions. This resource will help quickly onboard new developers who join your team (or serve as a reminder to yourself when you need to expand your current Braze implementation). 

## Maintenance {#maintenance}

> After handoff to your marketers, you will continue to serve as a resource for maintenance. You will pay attention to iOS and Android updates that might impact the Braze SDK and ensure that your third-party vendors are up to date. 

Reviewer: ```suggestion
After handoff to your marketers, you will continue to serve as a resource for maintenance. You will pay attention to iOS and Android updates that might impact the Braze SDK and make sure that your third-party vendors are up to date. 
```

Reviewer 2: Rachel the "ensure" queen 💅 ❤️

> You will track updates to the Braze platform via Braze's [GitHub][23]. Occasionally, your administrator will receive emails about urgent updates and bug fixes directly from Braze, as well. 

Reviewer: ```suggestion
You will track updates to the Braze platform via the Braze [GitHub][23]. Occasionally, your administrator will receive emails about urgent updates and bug fixes directly from Braze, as well. 
```
